Add center to Stencils.

evetion opened the pull request:
Fixes #57.

I had to split the stencil constructors to prevent ambiguity errors with StaticArrays. I do see that I miss some constructors now for nested arrays, the interface is quite large.

Currently I run into an error with the multi nested layers, I could use some help with that, even Cthulhu is not helping me. Could be a wrong constructor somewhere.

```julia
julia> mapstencil(A) do l
           sum(l.l1.b) - sum(l.l2.a)
       end
ERROR: ArgumentError: cannot convert a value to Union{} for assignment
Stacktrace:
  [1] convert(T::Type{Union{}}, x::Int64)
    @ Base ./essentials.jl:455
  [2] setindex!
    @ ./array.jl:994 [inlined]
  [3] setindex!
    @ ./multidimensional.jl:704 [inlined]
  [4] macro expansion
    @ ~/code/Stencils.jl/src/mapstencil.jl:112 [inlined]
  [5] cpu_mapstencil_kerneln0!
    @ ~/.julia/packages/KernelAbstractions/0r40T/src/macros.jl:293 [inlined]
  [6] cpu_mapstencil_kerneln0!(__ctx__::KernelAbstractions.CompilerMetadata{…}, f::var"#11#12", dest::Matrix{…}, source::StencilArray{…})
    @ Stencils ./none:0
  [7] __thread_run(tid::Int64, len::Int64, rem::Int64, obj::KernelAbstractions.Kernel{…}, ndrange::Nothing, iterspace::KernelAbstractions.NDIteration.NDRange{…}, args::Tuple{…}, dynamic::KernelAbstractions.NDIteration.DynamicCheck)
    @ KernelAbstractions ~/.julia/packages/KernelAbstractions/0r40T/src/cpu.jl:142
  [8] __run(obj::KernelAbstractions.Kernel{…}, ndrange::Nothing, iterspace::KernelAbstractions.NDIteration.NDRange{…}, args::Tuple{…}, dynamic::KernelAbstractions.NDIteration.DynamicCheck, static_threads::Bool)
    @ KernelAbstractions ~/.julia/packages/KernelAbstractions/0r40T/src/cpu.jl:109
  [9] (::KernelAbstractions.Kernel{…})(::Function, ::Vararg{…}; ndrange::Nothing, workgroupsize::Nothing)
    @ KernelAbstractions ~/.julia/packages/KernelAbstractions/0r40T/src/cpu.jl:44
 [10] Kernel
    @ ~/.julia/packages/KernelAbstractions/0r40T/src/cpu.jl:37 [inlined]
 [11] mapstencil!(::Function, ::Matrix{…}, ::StencilArray{…})
    @ Stencils ~/code/Stencils.jl/src/mapstencil.jl:102
 [12] mapstencil(::Function, ::StencilArray{Tuple{…}, 2, Int64, 2, Matrix{…}, Layered{…}, Remove{…}, Conditional})
    @ Stencils ~/code/Stencils.jl/src/mapstencil.jl:18
 [13] top-level scope
    @ REPL[9]:1
Some type information was truncated. Use `show(err)` to see complete types.

```

rafaqz commented:
Yes covering static arrays constructor ambiguity is a nightmare.

Can I be a huge pain and ask to swap the argument order to have center last, with a default of `nothing` ? Then this will be non-breaking

evetion commented:
I switched it around because some of the constructors  used splatting. I can try again, now that I've removed most of those.

If we want to allow nothing in all constructors, I'd have to discouple the T parameter (and add a new one for center?). Currently all empty constructors are backwards compatible.

evetion commented:
I've commented out the failing call described above. It boils down to that with a anonymous function, it can't infer the return type on multilayered stencils as

```julia
    return Base._return_type(f, Tuple{H,map(eltype, args)...})
```
becomes `Union{}`.

rafaqz commented:
> If we want to allow nothing in all constructors, I'd have to discouple the T parameter (and add a new one for center?). Currently all empty constructors are backwards compatible.

Ugh right yes maybe a break is better than an extra type parameter

evetion commented:
Setup `A = rand(3600,3600); moore = Moore(1)`

Before
```
julia> @benchmark mapstencil(mean, $moore, $A)
BenchmarkTools.Trial: 176 samples with 1 evaluation.
 Range (min … max):  26.678 ms … 68.851 ms  ┊ GC (min … max): 0.00% … 61.85%
 Time  (median):     28.205 ms              ┊ GC (median):    4.75%
 Time  (mean ± σ):   28.422 ms ±  3.124 ms  ┊ GC (mean ± σ):  5.59% ±  4.62%

                             ▁ ▃█▃▃▇▁                          
  ▃▃▃▁▃▁▁▁▁▃▃▃▁▁▃▁▄▃▃▃▅▅▅██▇▇██████████▇▆▄▃▁▁▃▁▁▁▁▁▃▁▁▁▃▁▁▃▁▃ ▃
  26.7 ms         Histogram: frequency by time        29.6 ms <

 Memory estimate: 98.88 MiB, allocs estimate: 76.
```
PR:
```
julia> @benchmark mapstencil(mean, $moore, $A)
BenchmarkTools.Trial: 159 samples with 1 evaluation.
 Range (min … max):  29.327 ms … 73.141 ms  ┊ GC (min … max): 0.00% … 60.09%
 Time  (median):     31.356 ms              ┊ GC (median):    4.39%
 Time  (mean ± σ):   31.520 ms ±  3.456 ms  ┊ GC (mean ± σ):  5.29% ±  4.73%

                       ▁▁             ▂  ▁█▂                   
  ▃▁▃▁▁▁▃▅▁▃▃▃▅▃▄▃██▆▃▇██▄▇▅▄▄▃▃▄▆▅▆▄▄█▆▇███▇▄▅▅▁▄▃▁▃▁▅▁▄▃▃▃▃ ▃
  29.3 ms         Histogram: frequency by time        32.9 ms <

 Memory estimate: 98.88 MiB, allocs estimate: 76.
```

11% slower, which is very close to extra center access for Moore{1} `8+1/8 = 1.125`.

I will add some tests, but otherwise this is good to go?

rafaqz commented:
Huh maybe we can make it optional to save the time. I was hoping the compiler could see it wasn't used but maybe not with array access

evetion commented:
That would complicate the code a bit, not only in terms of constructors, but the optional would require a new type parameter. Also, I'd like to get this in (and Geomorphometry depending on it released) for JuliaEO. But it's your package.

evetion commented:
Tinkering with it, you can cheat using the following patch, by not setting the center field, and not using it, for stencils that already include the center. It does does improve a ms for Window and similar stencils that already include the center. But frankly, not sure if it's worth it, given the complexity, and the fact the benchmark noise is around 10%...

```patch
diff --git a/src/array.jl b/src/array.jl
index 09759e3..2c6a5dd 100644
--- a/src/array.jl
+++ b/src/array.jl
@@ -29,7 +29,18 @@ Base.@propagate_inbounds stencil(hood::StencilOrLayered, A::AbstractStencilArray
     rebuild(hood, neighbors(hood, A, I), _center(hood, A, I))
 Base.@propagate_inbounds stencil(A::AbstractStencilArray) = A.stencil

-@inline _center(::Stencil, A::AbstractStencilArray, I::CartesianIndex) = A[I]
+@inline @generated function _center(hood::Stencil, A::AbstractStencilArray{<:Any,<:Any,T}, I::CartesianIndex) where {T}
+    for (I, O) in enumerate(offsets(hood))
+        if sum(abs.(O)) == 0
+            return :(_zero(T))
+        end
+    end
+    return :(A[I])
+end
+
+@inline _zero(::Type{Nothing}) = nothing
+@inline _zero(T) = zero(T)
+
 @inline _center(hood::Layered, A::AbstractStencilArray, I::CartesianIndex) = map(l -> _center(l, A, I), hood)

 """
diff --git a/src/stencil.jl b/src/stencil.jl
index 1e1dc26..129ce7b 100644
--- a/src/stencil.jl
+++ b/src/stencil.jl
@@ -102,7 +102,14 @@ Base.@propagate_inbounds indexat(hood::Stencil, center, i) = CartesianIndex(offs

 Return the value of the central cell in the stencil.
 """
-center(hood::Stencil) = getfield(hood, :center)
+@generated function center(hood::Stencil)
+    for (I, O) in enumerate(offsets(hood))
+        if sum(abs.(O)) == 0
+            return :(@inbounds neighbors(hood)[$I])
+        end
+    end
+    return :(getfield(hood, :center))
+end

 """
     distances(hood::Stencil)

```

rafaqz commented:
Yeah prob not worth it, the compiler will mostly figure it's the same read anyway if it's all inlined

rafaqz commented on src/stencil.jl:
```diff
@@ -97,6 +97,13 @@ end
+"""
+    center(x::Stencil)
+
+Return the value of the central cell in the stencil.
```
```suggestion
Return the value of the central cell a stencil is offset around. It may or may not be part of the stencil itself.
```

evetion commented:
Semantics, but this is not actually breaking if you look at the docstrings. The neighbor, center constructor is private in a sense.

rafaqz commented:
Hm maybe it's not

rafaqz commented:
Is this good to merge?

evetion commented:
Sure! Yes please ;)
